pkg/collector/printer/printer.go

```diff
@@ -0,0 +1,223 @@
+	return c
+}
+
+func NewWithFlags(app *kingpin.Application) types.Collector {
+	c := &collector{
+		printerInclude: app.Flag(
+			FlagPrinterInclude,
+			"Regular expression to match printers to collect metrics for",
+		).Default(ConfigDefaults.printerInclude).String(),
+		printerExclude: app.Flag(
+			FlagPrinterExclude,
+			"Regular expression to match printers to exclude",
+		).Default(ConfigDefaults.printerExclude).String(),
+	}
+	return c
+}
+
+func (c *collector) SetLogger(logger log.Logger) {
+	c.logger = log.With(logger, "collector", Name)
+}
+
+func (c *collector) Build() error {
+	c.printerJobStatus = prometheus.NewDesc(
+		prometheus.BuildFQName(types.Namespace, Name, "job_status"),
+		"A counter of printer jobs by status",
+		[]string{"printer", "status"},
+		nil,
+	)
+	c.printerStatus = prometheus.NewDesc(
+		prometheus.BuildFQName(types.Namespace, Name, "status"),
+		"Printer status",
+		[]string{"printer", "status"},
+		nil,
+	)
+	c.printerJobCount = prometheus.NewDesc(
+		prometheus.BuildFQName(types.Namespace, Name, "job_count"),
+		"Number of jobs processed by the printer since the last reset",
+		[]string{"printer"},
+		nil,
+	)
+
+	var err error
+	c.printerIncludePattern, err = regexp.Compile(fmt.Sprintf("^(?:%s)$", *c.printerInclude))
+	if err != nil {
+		return err
+	}
+	c.printerExcludePattern, err = regexp.Compile(fmt.Sprintf("^(?:%s)$", *c.printerExclude))
+	return err
+}
+
+func (c *collector) GetName() string { return Name }
+
+func (c *collector) GetPerfCounter() ([]string, error) { return []string{"Printer"}, nil }
+
+type win32_Printer struct {
+	Name                   string
+	Default                bool
+	PrinterStatus          uint16
+	JobCountSinceLastReset uint32
+}
+
+type win32_PrintJob struct {
+	Name   string
+	Status string
+}
+
+func (c *collector) Collect(_ *types.ScrapeContext, ch chan<- prometheus.Metric) error {
+	if err := c.collectPrinterStatus(ch); err != nil {
+		_ = level.Error(c.logger).Log("msg", "failed to collect printer status metrics", "err", err)
+		return err
+	}
+	if err := c.collectPrinterJobStatus(ch); err != nil {
+		_ = level.Error(c.logger).Log("msg", "failed to collect printer job status metrics", "err", err)
+		return err
+	}
+	return nil
+}
+
+func (c *collector) collectPrinterStatus(ch chan<- prometheus.Metric) error {
+	var printers []win32_Printer
+	q := wmi.QueryAll(&printers, c.logger)
+	if err := wmi.Query(q, &printers); err != nil {
+		return err
+	}
+
+	for _, printer := range printers {
+		if c.printerExcludePattern.MatchString(printer.Name) ||
+			!c.printerIncludePattern.MatchString(printer.Name) {
+			continue
+		}
+
+		printerStatus, ok := printerStatusMap[printer.PrinterStatus]
+		if !ok {
+			printerStatus = "Unknown"
+		}
+		ch <- prometheus.MustNewConstMetric(
+			c.printerStatus,
+			prometheus.GaugeValue,
+			1,
+			printer.Name,
+			printerStatus,
+		)
+		ch <- prometheus.MustNewConstMetric(
+			c.printerJobCount,
+			prometheus.GaugeValue,
```

Comment:
JobCounter is expect to increase only? If yes, it should be a Counter.

Reply:
good link!

According the linked docs, the value is documented as counter, too.

Reply:
Sure, I'll adjust it.

Reply:
Done ✅